providers/implementations/rands/drbg.c

```diff
@@ -924,5 +927,10 @@ int ossl_drbg_set_ctx_params(PROV_DRBG *drbg, const OSSL_PARAM params[])
     p = OSSL_PARAM_locate_const(params, OSSL_DRBG_PARAM_RESEED_TIME_INTERVAL);
     if (p != NULL && !OSSL_PARAM_get_time_t(p, &drbg->reseed_time_interval))
         return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_DRBG_PARAM_DETERMINISTIC);
```

Comment:
Is determinism something we want supported by all DRBGs?  Or just HMAC DRBG?

I kind of worry about it getting used by accident.

An alternative would be a separate deterministic HMAC DRBG implementation (which reuses just about all the HMAC DRBG code of course).

I don't have a feeling as to which approach is better.

Reply:
To separate it out you still end up with the following code....

````
int ossl_prov_drbg_generate_deterministic(PROV_DRBG *drbg,
                                          const unsigned char *adin,
                                          size_t adin_len,
                                          unsigned char *out, size_t outlen)
{
    if (!ossl_prov_is_running())
        return 0;

    if (drbg->state != EVP_RAND_STATE_READY) {
        /* try to recover from previous errors */
        rand_drbg_restart(drbg);

        if (drbg->state == EVP_RAND_STATE_ERROR) {
            ERR_raise(ERR_LIB_PROV, PROV_R_IN_ERROR_STATE);
            return 0;
        }
        if (drbg->state == EVP_RAND_STATE_UNINITIALISED) {
            ERR_raise(ERR_LIB_PROV, PROV_R_NOT_INSTANTIATED);
            return 0;
        }
    }
    if (outlen > drbg->max_request) {
        ERR_raise(ERR_LIB_PROV, PROV_R_REQUEST_TOO_LARGE_FOR_DRBG);
        return 0;
    }
    if (!drbg->generate(drbg, out, outlen, adin, adin_len)) {
        drbg->state = EVP_RAND_STATE_ERROR;
        ERR_raise(ERR_LIB_PROV, PROV_R_GENERATE_ERROR);
        return 0;
    }

    drbg->generate_counter++;

    return 1;
}
````

Reply:
and
````
int ossl_prov_drbg_instantiate_deterministic(
    PROV_DRBG *drbg,
    const unsigned char *entropy, size_t entropylen,
    const unsigned char *nonce, size_t noncelen)
{
    if (drbg->state != EVP_RAND_STATE_UNINITIALISED) {
        if (drbg->state == EVP_RAND_STATE_ERROR)
            ERR_raise(ERR_LIB_PROV, PROV_R_IN_ERROR_STATE);
        else
            ERR_raise(ERR_LIB_PROV, PROV_R_ALREADY_INSTANTIATED);
        goto end;
    }
    drbg->state = EVP_RAND_STATE_ERROR;

    if (!drbg->instantiate(drbg, entropy, entropylen, nonce, noncelen,
                           NULL, 0)) {
        ERR_raise(ERR_LIB_PROV, PROV_R_ERROR_INSTANTIATING_DRBG);
        goto end;
    }
    drbg->state = EVP_RAND_STATE_READY;
    drbg->generate_counter = 1;
 end:
    if (drbg->state == EVP_RAND_STATE_READY)
        return 1;
    return 0;
}
````

Reply:
i.e It needs to make sure a reseed doesnt happen..

Reply:
That's less painful than I thought it would be.

Reply:
So where would I stick this code in hmac_drbg or drbg.c?

Reply:
If its a seperate implementation it can also not have a parent and just set the entropy, nonce directly via setparams I guess

Reply:
hmac_drbg.c seems logical.  These two plus a new dispatch table should be enough.
I'd ignore things like parents and prediction_resistance.

Reply:
So you dont mind it doing the state checks in hmac_drbg?

Reply:
I guess it could go into it's own file and use the HMAC DRBG functions it needs.  Some renaming to be export friendly (ossl_XXX) would be required.